### DISCUSSION TITLE
Host & Device on separate cores?

### DISCUSSION THREAD

**Discussion by ATMakersBill** [hathach/tinyusb#1435] [General]
I am looking to use the new PIO Host mode to make HID Filters.  Specifically, we have filters that take non-standard (raw) HID and convert them to a composite Mouse + Keyboard that allows the old IntelliKeys keyboards to work for those who need them (primarily for those with cerebral palsy).  In addition, I'd like to make 'filtering" devices that take in (for example) a Mouse and expose the mouse out after a transform.  This would really help AT projects we're working on.

Our current filters use to MCUs: one playing the part of Host and one mimicking an HID Device.  So, for example, we use two Trinkets for this on the current IntelliKeys filters.

This makes for a nice separation of logic: the host side simply reads the reports and maintains a "current state" of which keys and buttons are pressed and the net motion of the mouse.  The device side polls and gets the updated state, sending appropriate reports out on the device connection.  This eliminates the problem if "stuck keys" where the key-up event is missed and solves the issues where one side (host/device) is significantly faster than the other.

So: I'd like to keep this approach in TinyUSB. However, in your concurrency doc, it says this is designed for a single core MCU.  Does that mean I can't run the device code in a separate core from the host code?  

Ideally, I'd like to have the main thread manage the host side, updating a shared "current state" structure that the second core of the RP2040 can read and send mouse & keyboard reports out the device side.

Is this a bad plan?  Will I have concurrency issues here?  Has anyone done anything like this?

Thanks
Bill

**Comment by hathach** [comment]
There should be no problems with each stack on its core and/or any combination of core + code as long as mutex is implemented correctly. just go ahead with your plan and file an issue if there is one

> So: I'd like to keep this approach in TinyUSB. However, in your concurrency doc, it says this is designed for a single core MCU. Does that mean I can't run the device code in a separate core from the host code?

can you paste that line here, so I could correct it.

**Reply by ATMakersBill** [reply]
Glad to hear that - here's the link (first paragraph)
https://docs.tinyusb.org/en/latest/reference/concurrency.html#concurrency

Thanks
Bill

**Reply by hathach** [reply]
thank I will remove that, preempted RTOS with multi-core is not too different in term of concurrency.

**Reply by Nikitarc** [reply]
Humm! Beware of multi-core with an RTOS.

There are two ways for an RTOS to use multiple cores.

In the case of Symmetric Multicore (SMP) a single RTOS manages the tasks on all the cores. In addition there is an API to assign tasks to particular cores (core affinity). In this case the mutex exists, and it corresponds to what is used by TinyUSB.

But most of the time in small systems an Asymmetric Multicore (AMP) is used. There is an independent RTOS running on each core, which only handles tasks running on that core. In this case, synchronization primitives between cores are usually limited to semaphore, message queues (and shared RAM) and mailboxes (an inter core mutex does not make sense in this architecture). And it's a different API from the ones that manage core task synchronization.

I'm not sure the mutex as TinyUSB uses it is suitable for AMP.
I think that for the targeted application a simple and common architecture can be used: a USB server. USB is managed by tasks on only 1 core, and the second core deposits requests in an inter-core message queue, and a dedicated task manages this queue and executes the requests (another queue is used for the other direction).

I think multi task and multi core are very different if an AMP OS is used. And the sentence in the doc is correct.
